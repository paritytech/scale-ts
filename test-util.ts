/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
