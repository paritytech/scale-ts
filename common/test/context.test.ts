import * as $ from "../../mod.ts"
import { assertEquals, assertThrowsSnapshot, testShape } from "../../test-util.ts"

Deno.test("ShapeVisitor", async (t) => {
  const visitor = new $.ShapeVisitor<string>()
  visitor
    .add($.u8, () => "$.u8")
    .add($.int, (_, signed, size) => `$.int(${signed}, ${size})`)
    .generic(<T>() => visitor.add($.array<T>, (_, $el) => `$.array(${visitor.visit($el)})`))

  assertEquals(visitor.visit($.array($.array($.u8))), "$.array($.array($.u8))")
  assertEquals(visitor.visit($.array($.u128)), "$.array($.int(false, 128))")
  await assertThrowsSnapshot(t, () => visitor.visit($.str))
})

type Graph = { label: string; to: Graph[] }

class GraphEncodeCtx {
  memo = new Map<Graph, number>()
}
class GraphDecodeCtx {
  memo: Graph[] = []
}

const $compactU32 = $.compact($.u32)
const $graph: $.Shape<Graph> = $.createShape({
  metadata: $.metadata("$graph"),
  staticSize: $compactU32.staticSize * 2 + $.str.staticSize,
  subEncode(buffer, value) {
    const ctx = buffer.context.get(GraphEncodeCtx)
    const key = ctx.memo.get(value)
    if (key != null) {
      return $compactU32.subEncode(buffer, key)
    }
    $compactU32.subEncode(buffer, ctx.memo.size)
    ctx.memo.set(value, ctx.memo.size)
    $.str.subEncode(buffer, value.label)
    $.array($graph).subEncode(buffer, value.to)
  },
  subDecode(buffer) {
    const ctx = buffer.context.get(GraphDecodeCtx)
    const key = $compactU32.subDecode(buffer)
    if (key < ctx.memo.length) {
      return ctx.memo[key]!
    }
    const graph: Graph = {
      label: $.str.subDecode(buffer),
      to: [],
    }
    ctx.memo.push(graph)
    graph.to = $.array($graph).subDecode(buffer)
    return graph
  },
  subAssert() {},
})

const a: Graph = { label: "a", to: [] }
const b: Graph = { label: "b", to: [] }
const c: Graph = { label: "c", to: [] }
const d: Graph = { label: "d", to: [] }
const e: Graph = { label: "e", to: [] }
const f: Graph = { label: "f", to: [] }
a.to = [b, c]
b.to = [d]
d.to = [a]
c.to = [a, e]
e.to = [a, c, f]
f.to = [a, b, c, d, e, f]

testShape($graph, { a, b, c, d, e, f })
