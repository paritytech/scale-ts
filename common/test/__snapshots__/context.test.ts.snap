export const snapshot = {};

snapshot[`ShapeVisitor 1`] = `Error: Unrecognized shape`;

snapshot[`\$graph a 1`] = `
00
04
61
08
04
04
62
04
08
04
64
04
00
0c
04
63
08
00
10
04
65
0c
00
0c
14
04
66
18
00
04
0c
08
10
14
`;

snapshot[`\$graph b 1`] = `
00
04
62
04
04
04
64
04
08
04
61
08
00
0c
04
63
08
08
10
04
65
0c
08
0c
14
04
66
18
08
00
0c
04
10
14
`;

snapshot[`\$graph c 1`] = `
00
04
63
08
04
04
61
08
08
04
62
04
0c
04
64
04
04
00
10
04
65
0c
04
00
14
04
66
18
04
08
00
0c
10
14
`;

snapshot[`\$graph d 1`] = `
00
04
64
04
04
04
61
08
08
04
62
04
00
0c
04
63
08
04
10
04
65
0c
04
0c
14
04
66
18
04
08
0c
00
10
14
`;

snapshot[`\$graph e 1`] = `
00
04
65
0c
04
04
61
08
08
04
62
04
0c
04
64
04
04
10
04
63
08
04
00
10
14
04
66
18
04
08
10
0c
00
14
`;

snapshot[`\$graph f 1`] = `
00
04
66
18
04
04
61
08
08
04
62
04
0c
04
64
04
04
10
04
63
08
04
14
04
65
0c
04
10
00
08
10
0c
14
00
`;
